# Analysis: How does Discover change a customer's purchasing behaviour?

#### Context: 
Previously, visitors of PCSG website may feel overwhelmed/at loss by the sheer amount of products listed. Having a product recommendation system aims to reduce the friction to navigation/purchase, thereby increasing sales.

#### Types of questions to answer:

1. How long does it take customers to buy something after using Discover?
    1. Completely new customers who have not made any purchase
    2. Existing customers who have made purchases before
    
    *Note: we do not have a "control group", because we do not know how long does it/used to take to get make their first product purchase*

#### Steps involved in formatting Discover User Data
Populate a table of emails, keeping the EARLIEST discover attempt of each customer 
Sample dataset of CSV file "discover_user_input_results.csv" is used in this analysis, which is obtained from the partial json file.

       **note to self: since this is a "partial" dataset, I will need to rerun this experiment with the full dataset**

In [7]:
import numpy as np
import pandas as pd
pd.set_option("display.max_columns", 100)

discover_data = pd.read_csv("discover_user_input_results.csv")
discover_data.head()


Unnamed: 0                   email  sensitivity     skinType  \
0         935  jeremy@paulaschoice.sg        False  Combination   
1         936  jeremy@paulaschoice.sg        False         Oily   
2         937   ck1411@singnet.com.sg        False  Combination   
3         938  jeremy@paulaschoice.sg        False         Oily   
4         939   starlites18@gmail.com        False  Combination   

                        timestamp       concern_0      concern_1  \
0  September 9th 2017, 1:56:08 pm     Dehydration     Sun Damage   
1  September 9th 2017, 1:58:21 pm            Acne       Wrinkles   
2  September 9th 2017, 2:38:56 pm  Enlarged Pores     Sun Damage   
3  September 9th 2017, 2:51:12 pm   Clogged Pores        Redness   
4  September 9th 2017, 2:59:07 pm  Enlarged Pores  Clogged Pores   

        concern_2       concern_3         concern_4 concern_5    concern_6  \
0             Men     Combination               NaN       NaN          NaN   
1             PIH             Men              Oily       NaN          NaN   
2        Wrinkles     Dehydration  Loss of Firmness  Dullness  Combination   
3  Uneven Texture  Enlarged Pores              Acne  Wrinkles  Dehydration   
4            Acne      Sun Damage    Uneven Texture   Redness  Combination   

    concern_7 concern_8 concern_9        concern_10 concern_11 concern_12  \
0         NaN       NaN       NaN               NaN        NaN        NaN   
1         NaN       NaN       NaN               NaN        NaN        NaN   
2         NaN       NaN       NaN               NaN        NaN        NaN   
3  Sun Damage       PIH  Dullness  Loss of Firmness        Men       Oily   
4         NaN       NaN       NaN               NaN        NaN        NaN   

  concern_13  result_0  result_1  result_2  result_3  result_4  result_5  \
0        NaN      7830      7780      7660      7740      7860      1720   
1        NaN      1150      7670      8720      7740      7870      6130   
2        NaN      7830      7780      7820      7740      7870      7760   
3        NaN      7830      7670      8720      7740      7870      7800   
4        NaN      6002      1350      2010      7740      7870      6130   

   result_6  result_7  result_8  result_9  
0      8740      2760      5800      7880  
1      8740      2760      5700      6240  
2      7690      2760      7900      7960  
3      8740      2760      5700      7730  
4      7690      2750      5700      7730

In [8]:
discover_first = discover_data.drop_duplicates(subset = ['email'], keep='first').copy()
discover_first['timestamp_proper'] = pd.to_datetime(discover_first['timestamp'].map(lambda x: x.replace('th', "")))
discover_first.drop('timestamp', axis=1,inplace=True)
discover_first.head()
# discover_first.to_csv('discover_first.csv')

Unnamed: 0                       email  sensitivity     skinType  \
0         935      jeremy@paulaschoice.sg        False  Combination   
2         937       ck1411@singnet.com.sg        False  Combination   
4         939       starlites18@gmail.com        False  Combination   
5         940  angelleexinyin@hotmail.com         True  Combination   
6         941     leong_lorna@hotmail.com         True         Oily   

        concern_0      concern_1       concern_2       concern_3  \
0     Dehydration     Sun Damage             Men     Combination   
2  Enlarged Pores     Sun Damage        Wrinkles     Dehydration   
4  Enlarged Pores  Clogged Pores            Acne      Sun Damage   
5            Acne  Clogged Pores  Uneven Texture  Enlarged Pores   
6   Clogged Pores           Acne             PIH  Enlarged Pores   

          concern_4    concern_5    concern_6    concern_7 concern_8  \
0               NaN          NaN          NaN          NaN       NaN   
2  Loss of Firmness     Dullness  Combination          NaN       NaN   
4    Uneven Texture      Redness  Combination          NaN       NaN   
5               PIH  Dehydration  Sensitivity  Combination       NaN   
6    Uneven Texture  Sensitivity         Oily          NaN       NaN   

  concern_9 concern_10 concern_11 concern_12 concern_13  result_0  result_1  \
0       NaN        NaN        NaN        NaN        NaN      7830      7780   
2       NaN        NaN        NaN        NaN        NaN      7830      7780   
4       NaN        NaN        NaN        NaN        NaN      6002      1350   
5       NaN        NaN        NaN        NaN        NaN      7830      1350   
6       NaN        NaN        NaN        NaN        NaN      6002      1350   

   result_2  result_3  result_4  result_5  result_6  result_7  result_8  \
0      7660      7740      7860      1720      8740      2760      5800   
2      7820      7740      7870      7760      7690      2760      7900   
4      2010      7740      7870      6130      7690      2750      5700   
5      6200      7740      7870      6130      7690      2750      5700   
6      6200      7740      7870      6130      7690      2750      5700   

   result_9    timestamp_proper  
0      7880 2017-09-09 13:56:08  
2      7960 2017-09-09 14:38:56  
4      7730 2017-09-09 14:59:07  
5      7730 2017-09-09 15:17:38  
6      7730 2017-09-09 15:38:11

### Steps involved in formatting Shopify Sales Data
1. Import Shopify sales_data (dated 2018-02-07) and perform basic formatting 
2. Match the emails available from Discover dataset to sales data, and create a new column called `"discover_first_date"` to indicate if, and when customer first used Discover
    1. Note: doing this excludes all the customers who used Discover but DID NOT make a purchase.
3. Calculate `'used_discover_already'` : whether a transaction is made BEFORE/AFTER the customer has tried Discover.
    1. Note: customers that have not tried Discover *at all* but made a purchase will also show "Not Yet"
    1. Perhaps those who have not tried Discover at all should have a third status ("Not at all")?
4. Calculate `'discover_sales_lead_time'`: time taken between the customer's first attempt of Discover and his/her current transaction.


In [9]:
sales_data = pd.read_csv("shopify_orders_export_20180207.csv", 
                         low_memory=False, 
                         parse_dates=['Paid at', 'Fulfilled at', 'Created at'])

sales_data_clean = sales_data.drop(sales_data.columns.to_series()[-11:-1], axis=1)
sales_data_clean.dropna(subset=['Email'], axis=0, inplace=True)

sales_data_clean['discover_first_date'] = sales_data_clean['Email'].map(discover_first.set_index('email')['timestamp_proper'])
sales_data_clean['used_discover_already'] = (sales_data_clean['Created at']> sales_data_clean['discover_first_date']).map({True: "Used Discover", False: "Not yet"})
sales_data_clean['discover_sales_lead_time'] = sales_data_clean['Created at'] - sales_data_clean['discover_first_date']

sales_data_clean.head()

Name                   Email Financial Status             Paid at  \
0  191569914712  gilly.glanville@me.com             paid 2018-02-08 03:59:00   
1  191569914712  gilly.glanville@me.com              NaN                 NaT   
2  191569914712  gilly.glanville@me.com              NaN                 NaT   
3  191569914712  gilly.glanville@me.com              NaN                 NaT   
4  191569914712  gilly.glanville@me.com              NaN                 NaT   

  Fulfillment Status        Fulfilled at Accepts Marketing Currency  Subtotal  \
0          fulfilled 2018-02-08 03:59:01               yes      SGD     290.0   
1                NaN                 NaT               NaN      NaN       NaN   
2                NaN                 NaT               NaN      NaN       NaN   
3                NaN                 NaT               NaN      NaN       NaN   
4                NaN                 NaT               NaN      NaN       NaN   

   Shipping  Taxes  Total     Discount Code  Discount Amount Shipping Method  \
0       0.0    0.0  290.0  5OFFe53b8c2fb9df              5.0             NaN   
1       NaN    NaN    NaN               NaN              NaN             NaN   
2       NaN    NaN    NaN               NaN              NaN             NaN   
3       NaN    NaN    NaN               NaN              NaN             NaN   
4       NaN    NaN    NaN               NaN              NaN             NaN   

           Created at  Lineitem quantity  \
0 2018-02-08 03:59:00                  1   
1 2018-02-08 03:59:00                  1   
2 2018-02-08 03:59:00                  1   
3 2018-02-08 03:59:00                  1   
4 2018-02-08 03:59:00                  1   

                                       Lineitem name  Lineitem price  \
0    Resist Skin Restoring Moisturizer SPF 50 - 60ml            48.0   
1  Resist Optimal Results Hydrating Cleanser - 19...            36.0   
2  Resist Advanced Smoothing Treatment 10% AHA (G...            55.0   
3                   Resist C15 Super Booster - 20 ml            68.0   
4     Clinical Ceramide-Enriched Firming Moisturizer            88.0   

   Lineitem compare at price Lineitem sku  Lineitem requires shipping  \
0                        0.0         7970                        True   
1                        0.0         7600                        True   
2                        0.0         7651                        True   
3                        0.0         7770                        True   
4                        0.0         2120                        True   

   Lineitem taxable Lineitem fulfillment status     Billing Name  \
0             False                   fulfilled  Gilly Glanville   
1             False                   fulfilled              NaN   
2              True                   fulfilled              NaN   
3             False                   fulfilled              NaN   
4             False                   fulfilled              NaN   

  Billing Street Billing Address1 Billing Address2 Billing Company  \
0            NaN              NaN              NaN             NaN   
1            NaN              NaN              NaN             NaN   
2            NaN              NaN              NaN             NaN   
3            NaN              NaN              NaN             NaN   
4            NaN              NaN              NaN             NaN   

  Billing City Billing Zip Billing Province Billing Country Billing Phone  \
0          NaN         NaN              NaN             NaN           NaN   
1          NaN         NaN              NaN             NaN           NaN   
2          NaN         NaN              NaN             NaN           NaN   
3          NaN         NaN              NaN             NaN           NaN   
4          NaN         NaN              NaN             NaN           NaN   

  Shipping Name Shipping Street Shipping Address1 Shipping Address2  \
0           NaN             NaN               NaN            

## Question 1-A: How long does it take for brand new customers to buy something after using Discover?
### First, identify (new) customers who had not made any purchase before Discover launch

1. Create a new column that indicates whether this customer has made any purchase before the launch of Discover (2017-09-09)
2. Create a table containing Emails, First Tried Discover Date and whether they existed before 2017-09-09 (Discover Launch)
    1. *Note: this EXCLUDES those customers who *have not tried Discover*, regardless of whether they have made a purchase before/after the launch of Discover*

1. Create two tables that include transactions before and on/after Discover launch (2017-09-09)

In [10]:
# Create two dataframes that include transactions before and on/after Discover launch (2017-09-09)
# (why?)
pre_discover_sales = sales_data_clean[sales_data_clean['Created at']< "2017-09-09"]
post_discover_sales = sales_data_clean[sales_data_clean['Created at']>= "2017-09-09"]

# Create a new dataframe from post_discover_sales
# that include only new users who took discover
# (why? what is the implication if I DON'T filter for pre-/post-Discover launch transactions?)

only_users_from_post_discover_sales = post_discover_sales[post_discover_sales['Email'].isin(discover_first['email'].unique())]
only_users_from_post_discover_sales

Name                        Email Financial Status  \
10       191569914707            jglyj82@gmail.com             paid   
11       191569914707            jglyj82@gmail.com              NaN   
22     191569914-1758          zarr.gyii@gmail.com             paid   
23     191569914-1758          zarr.gyii@gmail.com              NaN   
29     191569914-1755          ycobonpue@gmail.com             paid   
34       191569914693        karenkhor27@gmail.com             paid   
35       191569914693        karenkhor27@gmail.com              NaN   
36       191569914693        karenkhor27@gmail.com              NaN   
37       191569914693        karenkhor27@gmail.com              NaN   
42       191569914689         boazruth76@yahoo.com             paid   
43       191569914689         boazruth76@yahoo.com              NaN   
62       191569914682         bbggf.0901@gmail.com             paid   
63       191569914682         bbggf.0901@gmail.com              NaN   
70       191569914678      jeronblahblah@gmail.com             paid   
71       191569914678      jeronblahblah@gmail.com              NaN   
72       191569914678      jeronblahblah@gmail.com              NaN   
73       191569914678      jeronblahblah@gmail.com              NaN   
120      191569914656          ruienseah@gmail.com             paid   
121      191569914656          ruienseah@gmail.com              NaN   
122      191569914656          ruienseah@gmail.com              NaN   
124      191569914654      qingshuang111@gmail.com             paid   
125      191569914654      qingshuang111@gmail.com              NaN   
126      191569914653      qingshuang111@gmail.com             paid   
131      191569914649        roseleenlua@gmail.com             paid   
132      191569914649        roseleenlua@gmail.com              NaN   
134      191569914647     anisa.adhyanti@gmail.com             paid   
135      191569914647     anisa.adhyanti@gmail.com              NaN   
140      191569914645   parsheechandnani@gmail.com             paid   
144      191569914641            lukwern@gmail.com             paid   
145      191569914641            lukwern@gmail.com              NaN   
...               ...                          ...              ...   
10189     19156809852          ggwenny@hotmail.com             paid   
10190     19156809851          siyinglui@gmail.com             paid   
10191     19156809851          siyinglui@gmail.com              NaN   
10194     19156809849          renesis@outlook.com             paid   
10202     19156809845         a.krasocka@gmail.com             paid   
10203     19156809844          niko.seah@gmail.com             paid   
10204     19156809844          niko.seah@gmail.com              NaN   
10205     19156809844          niko.seah@gmail.com              NaN   
10212     19156809839           Stephtpy@gmail.com             paid   
10213     19156809838          anqi.la@hotmail.com             paid   
10214     19156809838          anqi.la@hotmail.com              NaN   
10215     19156809838          anqi.la@hotmail.com              NaN   
10216     19156809837        mirandalee88@live.com             paid   
10217     19156809837        mirandalee88@live.com              NaN   
10218     19156809837        mirandalee88@live.com              NaN   
10225     19156809834     lim.veronica21@yahoo.com             paid   
10226     19156809834     lim.veronica21@yahoo.com              NaN   
10237     19156809830       adlinadiyana@gmail.com             paid   
10238     19156809830       adlinadiyana@gmail.com              NaN   
10239     19156809829       dianaho87@yahoo.com.sg             paid   
10240     19156809829       dianaho87@yahoo.com.sg              NaN   
10241     19156809829       dianaho87@yahoo.com.sg              NaN   
10242     19156809829       dianaho87@yahoo.com.sg              NaN   
10243     19156809828        hayatiabdul@gmail.com             paid   
10244     19156809828        hayatiabdul@

In [37]:
sales_data_clean[sales_data_clean['Email'] == 'jovi.kau@gmail.com']

Name               Email Financial Status  \
605      191569914409  jovi.kau@gmail.com             paid   
606      191569914409  jovi.kau@gmail.com              NaN   
4085     191569912733  jovi.kau@gmail.com             paid   
6665     191569911539  jovi.kau@gmail.com             paid   
6666     191569911539  jovi.kau@gmail.com              NaN   
6667     191569911539  jovi.kau@gmail.com              NaN   
10160     19156809864  jovi.kau@gmail.com             paid   
11605     19156809177  jovi.kau@gmail.com             paid   
11606     19156809177  jovi.kau@gmail.com              NaN   
13558     19156808270  jovi.kau@gmail.com             paid   
13559     19156808270  jovi.kau@gmail.com              NaN   
16825  191568010-1139  jovi.kau@gmail.com             paid   
19778  191567910-1013  jovi.kau@gmail.com             paid   
19779  191567910-1013  jovi.kau@gmail.com              NaN   
21684     19156794624  jovi.kau@gmail.com             paid   
21685     19156794623  jovi.kau@gmail.com             paid   
22364     19156794305  jovi.kau@gmail.com             paid   
22812     19156794080  jovi.kau@gmail.com             paid   
23440     19156793795  jovi.kau@gmail.com             paid   
23441     19156793795  jovi.kau@gmail.com              NaN   
26297     19156792520  jovi.kau@gmail.com             paid   

                  Paid at Fulfillment Status        Fulfilled at  \
605   2018-01-30 01:51:08          fulfilled 2018-02-01 04:08:04   
606                   NaT                NaN                 NaT   
4085  2017-12-04 11:24:22          fulfilled 2017-12-04 15:39:46   
6665  2017-10-31 06:01:38          fulfilled 2017-11-01 07:32:16   
6666                  NaT                NaN                 NaT   
6667                  NaT                NaN                 NaT   
10160 2017-09-10 14:44:11          fulfilled 2017-09-11 06:52:51   
11605 2017-08-19 07:42:18          fulfilled 2017-08-21 15:26:25   
11606                 NaT                NaN                 NaT   
13558 2017-07-23 06:23:35          fulfilled 2017-07-24 08:02:16   
13559                 NaT                NaN                 NaT   
16825 2017-06-11 10:33:58          fulfilled 2017-06-11 10:33:58   
19778 2017-04-23 05:38:48          fulfilled 2017-04-23 05:38:47   
19779                 NaT                NaN                 NaT   
21684 2017-03-26 06:07:08          fulfilled 2017-03-26 06:07:08   
21685 2017-03-26 06:06:16          fulfilled 2017-03-26 06:06:17   
22364 2017-03-15 09:00:31          fulfilled 2017-03-16 06:52:05   
22812 2017-03-08 06:15:52          fulfilled 2017-03-08 09:12:18   
23440 2017-02-28 10:39:54          fulfilled 2017-02-28 10:39:54   
23441                 NaT                NaN                 NaT   
26297 2017-01-18 10:34:45          fulfilled 2017-01-18 10:34:45   

      Accepts Marketing Currency  Subtotal  Shipping  Taxes  Total  \
605                 yes      SGD      91.0       0.0    0.0   91.0   
606                 NaN      NaN       NaN       NaN    NaN    NaN   
4085                yes      SGD      82.0       0.0    0.0   82.0   
6665                yes      SGD     117.0       0.0    0.0  117.0   
6666                NaN      NaN       NaN       NaN    NaN    NaN   
6667                NaN      NaN       NaN       NaN    NaN    NaN   
10160               yes      SGD      82.0       0.0    0.0   82.0   
11605               yes      SGD      89.0       0.0    0.0   89.0   
11606               NaN      NaN       NaN       NaN    NaN    NaN   
13558               yes      SGD      82.0       0.0    0.0   82.0   
13559               NaN      NaN       NaN       NaN    NaN    NaN   
16825               yes      SGD      82.0       0.0    0.0   82.0   
19778               yes      SGD      89.0       0.0    0.0   89.0   
19779               NaN      NaN       NaN       NaN    NaN    NaN   
21684               yes      SGD      48.0       0.0    0.0   48.0   
21685               yes      SGD   

### Filter for transactions made within a day of using Discover (for the first time)
Into a table called `same_day_purchase`

In [43]:
'''
comparison_table = sales_data_clean[['Email', 'Created at', 'discover_first_date', 'used_discover_already']]
comparison_table['discover_sales_lead_time'] = comparison_table['Created at'] - comparison_table['discover_first_date']
'''
# Filter for transactions that were made within a day of attempting Discover
same_day_purchase = sales_data_clean[(sales_data_clean['discover_sales_lead_time'] <= "1 days") & (sales_data_clean['discover_sales_lead_time'] > "0 days")]
same_day_purchases_count = len(same_day_purchase['Email'].unique())

# Perform a count and percentage of customer analysis 
post_discover_launch_customer_count = len(post_discover_sales['Email'].unique())
percentage_same_day_purchase = same_day_purchases_count/post_discover_launch_customer_count
print("same_day_purchases_count:", same_day_purchases_count)
print("post_discover_launch_customer_count:", post_discover_launch_customer_count)
print("percentage_same_day_purchase: {:.2f}%".format(100*percentage_same_day_purchase))

# same_day_purchase.to_excel("same_day_purchase.xlsx")



same_day_purchases_count: 335
post_discover_launch_customer_count: 2847
percentage_same_day_purchase: 11.77%


# 12% of customers made a purchase within a day after using Discover!

## How long does it take for brand new customers to buy something after using Discover?
### Brand new customers = customers who had not purchased anything pre-Discover launch
There will be two types of post-launch new customers: 
1. Those who made their first purchase after trying out Discover
2. Those who made their first purchase WITHOUT trying out Discover 
    2. (either because they didn't bother, or didn't know of its existence)
3. Those who made their first purchase WITHOUT trying out Discover (i.e. point 2, but subsequently tried Discover)
    3. This group is currently ignored

In [53]:
# Create a new dataframe containing UNIQUE emails from post_discover_sales
post_launch_emails = pd.DataFrame(post_discover_sales['Email'].unique(), 
                                  columns=["Post Launch Emails"])

# Test if these emails are also found in pre_discover_sales email list
post_launch_emails['Exist before launch?'] = post_launch_emails.isin(pre_discover_sales['Email'].unique())

# Merge (aka Join) Discover Data with emails in post_launch_emails
# and create columns to indicate whether and when did the email tried discover
post_launch_emails = post_launch_emails.merge(right=discover_first[['email', 'timestamp_proper']], 
                                              left_on='Post Launch Emails', 
                                              right_on='email')

# Delete the extra email column created after joining
post_launch_emails = post_launch_emails.drop('email', axis=1)

# Rename the column into something more elucidatory
post_launch_emails = post_launch_emails.rename(index=str, columns={"timestamp_proper": "First Tried Discover"})

# get first transaction date of each of the "new" customers
# Remember that sales_data is reverse-chronologically indexed
post_launch_first_purchase = post_discover_sales[['Email', 'Created at']].drop_duplicates(subset='Email', keep='last')
print("Number of rows:", len(post_launch_first_purchase))
post_launch_first_purchase

post_launch_emails = post_launch_emails.merge(post_launch_first_purchase, 
                                              left_on="Post Launch Emails", 
                                              right_on="Email")
post_launch_emails = post_launch_emails.drop('Email', axis=1).rename(index=str, 
                                                                     columns={"Created at": "First Transaction Post Launch"})

post_launch_emails['Time To Buy'] = post_launch_emails['First Transaction Post Launch'] - post_launch_emails['First Tried Discover']

# Filter for purchases made in 1 day (if necessary)
# post_launch_emails = post_launch_emails[(post_launch_emails['Time To Buy']> "0 days") & (post_launch_emails['Exist before launch?'] == False)]

# Lead time to purchase for new customers after taking Discover
print(post_launch_emails[(post_launch_emails['Time To Buy'] > '0 days') & 
                         (post_launch_emails['Exist before launch?'] == False)]['Time To Buy'].describe())
print('Top 85% of time to buy:', post_launch_emails['Time To Buy'].quantile(0.85))

post_launch_emails

Number of rows: 2847
count                        346
mean     10 days 03:05:21.540462
std      21 days 16:32:44.176929
min              0 days 00:01:22
25%       0 days 00:19:19.500000
50%       0 days 09:24:47.500000
75%       9 days 00:29:25.250000
max            139 days 05:01:07
Name: Time To Buy, dtype: object
Top 85% of time to buy: 25 days 16:44:05.549999


Post Launch Emails  Exist before launch? First Tried Discover  \
0               jglyj82@gmail.com                 False  2018-02-07 12:54:54   
1             zarr.gyii@gmail.com                  True  2017-09-26 03:45:14   
2             ycobonpue@gmail.com                  True  2017-10-31 06:06:57   
3           karenkhor27@gmail.com                  True  2017-09-13 02:03:15   
4            boazruth76@yahoo.com                 False  2018-01-24 09:57:32   
5            bbggf.0901@gmail.com                  True  2017-10-27 17:38:14   
6         jeronblahblah@gmail.com                  True  2017-09-17 05:58:14   
7             ruienseah@gmail.com                 False  2018-02-06 08:16:14   
8         qingshuang111@gmail.com                 False  2017-10-01 15:37:11   
9           roseleenlua@gmail.com                  True  2018-01-16 05:28:25   
10       anisa.adhyanti@gmail.com                  True  2017-09-10 10:00:33   
11     parsheechandnani@gmail.com                 False  2017-10-31 06:32:51   
12              lukwern@gmail.com                 False  2017-10-14 09:05:21   
13   lohhuijuanrachel@hotmail.com                 False  2018-01-27 19:18:42   
14     shireengill.1991@gmail.com                  True  2018-02-05 14:31:05   
15            syshuka@hotmail.com                 False  2018-01-29 22:18:16   
16       imobviouslynot@gmail.com                 False  2018-02-04 19:43:12   
17            lyejiahui@gmail.com                  True  2017-09-09 16:07:52   
18            mabolight@gmail.com                 False  2017-10-21 04:13:03   
19          mirandalee88@live.com                  True  2017-11-01 02:48:44   
20             yyanti22@yahoo.com                  True  2017-10-07 14:34:53   
21      hazelwongshihui@gmail.com                  True  2017-09-10 06:38:23   
22    somu_luvhrithik@hotmail.com                  True  2017-10-06 05:00:39   
23              manasa@luxola.com                  True  2017-10-07 07:47:14   
24               pkly58@gmail.com                  True  2017-10-31 07:44:52   
25               tnxw@hotmail.com                  True  2017-10-31 08:24:13   
26        jessiejiaqiye@gmail.com                  True  2017-10-23 10:32:45   
27      tranlethuthuy94@gmail.com                 False  2017-10-09 05:03:47   
28          yenwei_chea@yahoo.com                  True  2018-02-03 14:16:00   
29   amanda_wong2003@yahoo.com.sg                 False  2018-02-03 01:17:14   
..                            ...                   ...                  ...   
698       giovanna.dioni@yahoo.it                  True  2017-09-15 08:54:43   
699         linmeinmein@gmail.com                 False  2017-09-16 05:52:45   
700          laipeiwong@gmail.com                 False  2017-09-10 04:23:57   
701         tram.hien27@gmail.com                 False  2017-09-16 05:04:43   
702           gyang13@hotmail.com                  True  2017-09-10 07:38:37   
703           epsh.1015@gmail.com                  True  2017-09-11 17:09:53   
704         plastickers@gmail.com                  True  2017-09-15 16:15:31   
705        syahirah17@hotmail.com                 False  2017-09-14 16:10:03   
706    social_fantasy@hotmail.com                 False  2017-09-15 02:46:01   
707      twofinerascals@gmail.com                 False  2017-09-14 17:54:55   
708           dionislim@gmail.com                 False  2017-09-14 14:31:55   
709         abc-kevin@live.com.sg                  True  2017-09-28 23:06:16   
710          kly0606@yahoo.com.sg                 False  2017-09-13 18:31:04   
711               mei_yin@live.cn                 False  2017-09-11 01:58:55   
712      joansoonjialin@gmail.com                  True  2017-09-13 18:33:47   
713         lyelye241@hotmail.com                 False  2017-09-13 03:03:12   
714             aprxlim@gmail.com                  True  2017-09-13 02:57:46   
715         xinyan_11@hotmail.com                 False  2017-10-31 05:52:54   
716          Ba

## 50% of our new customers made a purchase within 10 hours after trying out Discover!


In [38]:
# post_launch_emails[post_launch_emails['Post Launch Emails'] == 'jovi.kau@gmail.com']

Post Launch Emails  Exist before launch? First Tried Discover  \
66  jovi.kau@gmail.com                  True  2017-10-31 05:49:38   

   First Transaction Post Launch        Time To Buy  
66           2017-09-10 14:44:10 -51 days +08:54:32

From the calculation above, we know that: 
1. there are 2847 first-time purchases since Discover was launched.
2. on average, it takes almost **7 days** for a customer to make a purchase after using Discover
3. 50% of transactions occured since Discover was launched occured within **5 hours** of trying out Discover

#### Obtain a list of unique emails of NEW customers who made their first purchase *without* taking Discover

In [40]:
new_cust_without_discover = post_launch_emails[(post_launch_emails['Time To Buy'] < pd.Timedelta('00:00:00')) &
                                               (post_launch_emails['Exist before launch?'] == False)]
new_cust_without_discover

Post Launch Emails  Exist before launch? First Tried Discover  \
4           boazruth76@yahoo.com                 False  2018-01-24 09:57:32   
11    parsheechandnani@gmail.com                 False  2017-10-31 06:32:51   
34         trac_tan@yahoo.com.sg                 False  2018-01-08 16:16:34   
43          weejingshan@mail.com                 False  2017-10-30 17:42:25   
48         jeenaujames@gmail.com                 False  2018-01-23 08:07:22   
87         eeliew@rocketmail.com                 False  2017-10-24 13:45:18   
88           lilyloh26@gmail.com                 False  2018-01-11 13:26:06   
93         norainialwi@yahoo.com                 False  2018-01-03 00:20:53   
96        huiting.chen@yahoo.com                 False  2018-02-04 09:56:00   
102         liewjas@yahoo.com.sg                 False  2018-01-23 07:03:59   
103     kate.cookson@clydeco.com                 False  2018-01-23 04:43:34   
114           ngabizin@gmail.com                 False  2018-01-19 07:34:23   
125          siyinglui@gmail.com                 False  2017-09-21 07:57:37   
130       gillianteo06@gmail.com                 False  2018-02-17 16:13:32   
151        dianachan89@gmail.com                 False  2017-10-09 11:49:30   
163           mindylor@gmail.com                 False  2017-12-27 02:34:30   
169         lijuan.tlj@gmail.com                 False  2018-01-05 05:49:56   
171      abrizah.haris@gmail.com                 False  2017-10-18 01:13:38   
181  chloe.yuting.wang@gmail.com                 False  2017-10-15 08:04:19   
189           yshflora@gmail.com                 False  2017-12-31 13:09:19   
207    angeliaghtan@yahoo.com.sg                 False  2017-12-27 03:12:53   
210         poppingtart@live.com                 False  2017-10-08 13:12:54   
223        emailevanak@gmail.com                 False  2018-01-27 02:45:21   
250       cassandra_yz@yahoo.com                 False  2018-01-23 06:58:12   
274            dwftr24@gmail.com                 False  2017-09-14 14:35:08   
276      limhweezhi.13@gmail.com                 False  2018-01-19 15:45:52   
296        johanarosli@gmail.com                 False  2017-09-20 13:20:59   
305       rachel.sng@hotmail.com                 False  2018-01-04 09:03:48   
322     sahanaprabhakar@live.com                 False  2017-11-27 05:44:39   
335       julianne.toh@gmail.com                 False  2017-11-28 14:36:18   
352       vitasin.7889@gmail.com                 False  2017-12-07 00:59:09   
355       hongloan.sam@gmail.com                 False  2017-09-17 06:58:55   
369     yaminmomo.ymmm@gmail.com                 False  2017-11-18 10:44:43   
373        missydevlin@gmail.com                 False  2017-11-13 07:38:53   
376         itiong.tip@gmail.com                 False  2018-02-16 04:42:58   
394          shy_karen@yahoo.com                 False  2018-01-16 15:26:47   
415           rouhui90@gmail.com                 False  2017-10-30 14:58:44   
426    weihao_toh@rocketmail.com                 False  2017-10-31 09:22:57   
523           tavia225@gmail.com                 False  2017-12-29 06:25:32   
546          wyfsuperb@gmail.com                 False  2017-11-04 03:51:11   
561         sock_mei@hotmail.com                 False  2017-11-06 09:16:05   
604       aeiou1280@yahoo.com.sg                 False  2017-10-06 08:26:37   
607       isabel.tanyx@gmail.com                 False  2018-02-06 06:08:56   
633   athena_hades72@hotmail.com                 False  2017-10-08 12:23:54   
656          ryanjh977@gmail.com                 False  2017-10-04 15:11:39   
661           leo75ren@yahoo.com                 False  2017-09-24 11:22:59   
678            bearyhk@gmail.com                 False  2017-10-31 08:33:15   
683       xue_qing31@hotmail.com                 False  2017-09-20 05:46:41   
715        xinyan_11@hotmail.com                 False  2017-10-31 05:52:54   
720   jessicanomanbhoy@yahoo.com                 False  2017-09-

In [36]:

# Obtain a list of new customers who had made their first purchase post-launch without using Discover
new_cust_without_discover = post_launch_emails[(post_launch_emails['Time To Buy'] < pd.Timedelta('00:00:00')) & 
                                               (post_launch_emails['Exist before launch?'] == False)] 
                                                #['Post Launch Emails'].unique()

post_discover_sales[~post_discover_sales['Email'].isin(new_cust_without_discover['Post Launch Emails'].unique())]

# post_launch_emails was already defined earlier, and have not been altered in any way
first_sales_without_discover = post_discover_sales[~post_launch_emails.isin(pre_discover_sales['Email'].unique())]
first_sales_without_discover = first_sales_without_discover.dropna()
first_sales_without_discover

2              ycobonpue@gmail.com
4             boazruth76@yahoo.com
11      parsheechandnani@gmail.com
14      shireengill.1991@gmail.com
19           mirandalee88@live.com
23               manasa@luxola.com
26         jessiejiaqiye@gmail.com
33        lim.veronica21@yahoo.com
34           trac_tan@yahoo.com.sg
40          rinadarman08@gmail.com
41            ancarujoiu@yahoo.com
43            weejingshan@mail.com
45         Jennyann57@yahoo.com.sg
48           jeenaujames@gmail.com
49           ana_adl27@hotmail.com
53        faithyeoh.gw95@gmail.com
61            Xjxmandy@hotmail.com
66              jovi.kau@gmail.com
74            zhujin0928@gmail.com
75          yashoda_20@hotmail.com
79            jannawqe@hotmail.com
87           eeliew@rocketmail.com
88             lilyloh26@gmail.com
89         ivyaustasia@hotmail.com
93           norainialwi@yahoo.com
95           hayatiabdul@gmail.com
96          huiting.chen@yahoo.com
98           artspassion@gmail.com
101         seankais

In [33]:
type(new_cust_without_discover['Time To Buy'][0])

pandas._libs.tslib.Timedelta

In [22]:
new_cust_without_discover = post_launch_emails[post_launch_emails['Time To Buy'] < pd.Timedelta('00:00:00')]
new_cust_without_discover

Post Launch Emails  Exist before launch? First Tried Discover  \
2            ycobonpue@gmail.com                  True  2017-10-31 06:06:57   
4           boazruth76@yahoo.com                 False  2018-01-24 09:57:32   
11    parsheechandnani@gmail.com                 False  2017-10-31 06:32:51   
14    shireengill.1991@gmail.com                  True  2018-02-05 14:31:05   
19         mirandalee88@live.com                  True  2017-11-01 02:48:44   
23             manasa@luxola.com                  True  2017-10-07 07:47:14   
26       jessiejiaqiye@gmail.com                  True  2017-10-23 10:32:45   
33      lim.veronica21@yahoo.com                  True  2017-09-10 06:43:12   
34         trac_tan@yahoo.com.sg                 False  2018-01-08 16:16:34   
40        rinadarman08@gmail.com                  True  2017-10-31 13:36:28   
41          ancarujoiu@yahoo.com                  True  2017-11-01 03:16:27   
43          weejingshan@mail.com                 False  2017-10-30 17:42:25   
45       Jennyann57@yahoo.com.sg                  True  2017-10-26 07:13:18   
48         jeenaujames@gmail.com                 False  2018-01-23 08:07:22   
49         ana_adl27@hotmail.com                  True  2017-09-28 13:24:08   
53      faithyeoh.gw95@gmail.com                  True  2017-09-12 08:53:49   
61          Xjxmandy@hotmail.com                  True  2017-10-25 00:53:36   
66            jovi.kau@gmail.com                  True  2017-10-31 05:49:38   
74          zhujin0928@gmail.com                  True  2017-11-11 01:05:44   
75        yashoda_20@hotmail.com                  True  2017-11-11 03:26:34   
79          jannawqe@hotmail.com                  True  2017-12-31 18:49:57   
87         eeliew@rocketmail.com                 False  2017-10-24 13:45:18   
88           lilyloh26@gmail.com                 False  2018-01-11 13:26:06   
89       ivyaustasia@hotmail.com                  True  2017-10-31 15:35:00   
93         norainialwi@yahoo.com                 False  2018-01-03 00:20:53   
95         hayatiabdul@gmail.com                  True  2017-10-10 01:54:19   
96        huiting.chen@yahoo.com                 False  2018-02-04 09:56:00   
98         artspassion@gmail.com                  True  2017-10-21 12:41:17   
101       seankaiser11@gmail.com                  True  2018-01-15 05:15:48   
102         liewjas@yahoo.com.sg                 False  2018-01-23 07:03:59   
..                           ...                   ...                  ...   
425       hyongzhi88@hotmail.com                  True  2017-10-31 09:07:41   
426    weihao_toh@rocketmail.com                 False  2017-10-31 09:22:57   
465     jocelyn.ng.foo@gmail.com                  True  2017-10-31 06:39:15   
466    freakdoomdragon@yahoo.com                  True  2017-11-11 05:30:06   
512       actdrive.ang@gmail.com                  True  2017-10-31 05:53:21   
523           tavia225@gmail.com                 False  2017-12-29 06:25:32   
544      nzafirah_97@hotmail.com                  True  2017-10-15 06:36:01   
546          wyfsuperb@gmail.com                 False  2017-11-04 03:51:11   
558         bungeegirl@gmail.com                  True  2017-10-10 02:31:59   
561         sock_mei@hotmail.com                 False  2017-11-06 09:16:05   
601      nessya.andina@gmail.com                  True  2018-02-06 05:07:12   
604       aeiou1280@yahoo.com.sg                 False  2017-10-06 08:26:37   
607       isabel.tanyx@gmail.com                 False  2018-02-06 06:08:56   
633   athena_hades72@hotmail.com                 False  2017-10-08 12:23:54   
638     demona.pastime@gmail.com                  True  2017-11-29 06:13:34   
656          ryanjh977@gmail.com                 False  2017-10-04 15:11:39   
661           leo75ren@yahoo.com                 False  2017-09-24 11:22:59   
665      lolit.bendico@gmail.com                  True  2017-11-01 09:29:53   
667          verosutji@yahoo.com                  True  2017-11-

## Average Order Value of _New_ Customers who made their first purchase after taking Discover

In [22]:
new_discover_sale_transaction = post_discover_sales[post_discover_sales['Email'].isin(new_customers_test['Post Launch Emails'].unique())]
# new_discover_sale_transaction.groupby(by=['Email', 'Created at'])['Total'].sum()
new_discover_sale_transaction['Total'].mean()

92.17494824016563

## Average Order Value of _New_ Customers who made their first purchase _without_ taking Discover

In [23]:
new_non_discover_sale_transaction = post_discover_sales[~post_discover_sales['Email'].isin(new_customers_test['Post Launch Emails'].unique())]
new_non_discover_sale_transaction['Total'].mean()

105.15647869674186

# Comparison

In [24]:
table_to_compare = pivot_discover_first
source_table_to_compare = sales_data_clean

In [25]:
multiindex_output = table_to_compare.set_index(keys=['email', 'SKU'])
multiindex_output

step_no
email                 SKU           
007lavender@gmail.com 6002  result_0
                      7670  result_1
                      6200  result_2
                      7740  result_3
                      7870  result_4
                      6130  result_5
                      7690  result_6
                      2750  result_7
                      5700  result_8
                      6240  result_9
1060138073@qq.com     1150  result_0
                      1350  result_1
                      2010  result_2
                      7740  result_3
                      7770  result_4
                      6130  result_5
                      3400  result_6
                      2750  result_7
                      5700  result_8
                      7730  result_9
1115006804@qq.com     1150  result_0
                      8510  result_1
                      6200  result_2
                      7740  result_3
                      7980  result_4
                      6130  result_5
                      3400  result_6
                      2750  result_7
                      5700  result_8
                      7730  result_9
...                              ...
zwying@gmail.com      6002  result_0
                      7670  result_1
                      6200  result_2
                      7740  result_3
                      7870  result_4
                      7800  result_5
                      7690  result_6
                      2750  result_7
                      5700  result_8
                      7730  result_9
zyan0981@gmail.com    1000  result_0
                      1250  result_1
                      2010  result_2
                      2130  result_3
                      7840  result_4
                      7970  result_5
                      7810  result_6
                      2800  result_7
                      5000  result_8
                      5200  result_9
NaN                   7830  result_0
                      1350  result_1
                      6200  result_2
                      7740  result_3
                      7860  result_4
                      6130  result_5
                      7690  result_6
                      2750  result_7
                      5700  result_8
                      7730  result_9

[24220 rows x 1 columns]

In [26]:
check_presence = multiindex_output.isin(source_table_to_compare[['Email', 'Lineitem sku']])
check_presence[check_presence['step_no'] == True]

Empty DataFrame
Columns: [step_no]
Index: []

### Following code aims to examine the proportion of SKUs *actually* recommended to the customers
#### This helps to examine the trend of skin type (better to get from raw data for this)
"unpivot" ("melt" in Pandas term) in discover_first[result_SKUs].
### The objective is to allow for easier indexing to search for SKUs in sales_data_clean

In [10]:
# pivot_discover_first = discover_first.unstack(index='email', columns = ['result_0', 'result_1',
#       'result_2', 'result_3', 'result_4', 'result_5', 'result_6', 'result_7',
#       'result_8', 'result_9'])
pivot_discover_first = pd.melt(discover_first, 
                               id_vars='email',
                               value_vars=list(discover_first.columns[-11:-1]),
                               var_name='step_no',
                               value_name= 'SKU')

In [11]:
pivot_discover_first.sort_values(['email', 'step_no'], inplace=True)
# pivot_discover_first.to_excel("pivot_discover_first.xlsx")
# pivot_discover_first
discover_first.columns[-10:-1]
discover_first[discover_first['email'] == "yuleandra21@gmail.com"]
# same_day_purchase.groupby(by=['Email','Lineitem sku'])['Lineitem sku'].count()
#pivot_discover_first['Purchased?'] = pivot_discover_first['email'].map(same_day_purchase.set_index(['Email', 'Lineitem sku'])[])

'''Faced issues creating a frequency plot for SKUs under each email based on sales data
Abandoned using jupyter, tried on excel instead, also gave up subsequently
(Might need to use for loop with two filter criteria, but could not figure out how)'''